src/v/kafka/server/handlers/fetch.cc

```diff
@@ -302,27 +371,30 @@ static ss::future<read_result> read_from_partition(
       0,
       config.max_bytes,
       kafka_read_priority(),
-      raft::data_batch_type,
+      std::nullopt,
       std::nullopt,
       std::nullopt);
 
     reader_config.strict_max_bytes = config.strict_max_bytes;
     return pw.make_reader(reader_config)
       .then([hw, lso, foreign_read, deadline](model::record_batch_reader rdr) {
-          // if we are on remote core, we MUST use foreign record batch reader.
-          if (foreign_read) {
-              return model::consume_reader_to_memory(
-                       std::move(rdr), deadline.value_or(model::no_timeout))
-                .then([hw, lso](ss::circular_buffer<model::record_batch> data) {
-                    return read_result(
-                      model::make_foreign_memory_record_batch_reader(
-                        std::move(data)),
-                      hw,
-                      lso);
-                });
-          }
-          return ss::make_ready_future<read_result>(
-            read_result(std::move(rdr), hw, lso));
+          return model::transform_reader_to_memory(
+                   std::move(rdr),
+                   deadline.value_or(model::no_timeout),
+                   adapt_fetch_batch)
+            .then([foreign_read](
+                    ss::circular_buffer<model::record_batch> data) {
+                // if we are on remote core, we MUST use foreign record batch
+                // reader.
+                if (foreign_read) {
+                    return model::make_foreign_memory_record_batch_reader(
+                      std::move(data));
+                }
+                return model::make_memory_record_batch_reader(std::move(data));
+            })
+            .then([hw, lso](model::record_batch_reader rdr) {
+                return read_result(std::move(rdr), hw, lso);
```

Comment:
does this have any important memory implications? before this may have been a log reader, but now it is always a memory-backed reader.

Reply:
It does not, the only path we didn't consume batches to memory was when reader and connection were coming from the same core. The only difference introduced by this code is that even for the same core we consume batches to memory